libtenzir/include/tenzir/multi_series_builder.hpp

```diff
@@ -0,0 +1,446 @@
+  using record_generator = detail::multi_series_builder::record_generator;
+
+  /// @returns a vector of all currently finished series
+  [[nodiscard("The result of a flush must be handled")]]
+  auto yield_ready() -> std::vector<series>;
+  /// @returns a vector of all currently finished series
+  [[nodiscard("The result of a flush must be handled")]]
+  auto yield_ready_as_table_slice() -> std::vector<table_slice>;
+
+  /// adds a record to the currently active builder
+  [[nodiscard]] auto record() -> record_generator;
+
+  /// drops the last event from active builder
+  void remove_last();
+
+  [[nodiscard("The result of a flush must be handled")]]
+  auto finalize() -> std::vector<series>;
+  [[nodiscard("The result of a flush must be handled")]]
+  auto finalize_as_table_slice() -> std::vector<table_slice>;
+
+  // this policy will merge all events into a single schema
+  struct policy_merge {
+    static constexpr std::string_view name = "merge";
+    // a schema name to seed with. If this is given
+    std::string seed_schema = {};
+    bool reset_on_yield = false;
+
+    auto friend inspect(auto& f, policy_merge& x) -> bool {
+      return f.object(x).fields(f.field("seed_schema", x.seed_schema),
+                                f.field("reset_on_yield", x.reset_on_yield));
+    }
+  };
+
+  // this policy will keep all schemas in separate batches
+  struct policy_precise {
+    static constexpr std::string_view name = "precise";
+    // If this is given, all resulting events will have exactly this schema
+    // * all fields in the schema but not in the event will be null
+    std::string seed_schema = {};
+
+    auto friend inspect(auto& f, policy_precise& x) -> bool {
+      return f.object(x).fields(f.field("seed_schema", x.seed_schema));
+    }
+  };
+
+  // this policy will keep all schemas in batches according to selector
+  struct policy_selector {
+    static constexpr std::string_view name = "selector";
+    // the field name to use for selection
+    std::string field_name;
+    // a naming prefix, doing the following transformation on the name:
+    // selector("event_type", "suricata")
+    // => {"event_type": "flow"}
+    // => "suricata.flow"
+    std::optional<std::string> naming_prefix = std::nullopt;
+    bool unique_selector = false;
+
+    auto friend inspect(auto& f, policy_selector& x) -> bool {
+      return f.object(x).fields(f.field("field_name", x.field_name),
+                                f.field("naming_prefix", x.naming_prefix),
+                                f.field("unique_selector", x.unique_selector));
+    }
+  };
+
+  // TODO the monostate alternative only exists because of an issue when
+  // compiling with GCC-12 in the CI
+  using policy_type = tenzir::variant<std::monostate, policy_merge,
+                                      policy_precise, policy_selector>;
+
+  struct settings_type {
+    // the default name given to a schema
+    std::string parser_name = "tenzir.unknown";
```

Comment:
We discussed in a call that this is not the right name for this, and that it should not be prepended to diagnostic messages.